File: solutions/0257/1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<>();
        traverse(root, "", result);
        return result;
    }

    private void traverse(TreeNode root, String s, List<String> result) {
        if (root == null) {
            return;
        }
        String ss;
        if (s.isEmpty()) {
            ss = String.valueOf(root.val);
        } else {
            ss = s + "->" + root.val;
        }
        if (root.left == null && root.right == null) {
            result.add(ss);
        }
        traverse(root.left, ss, result);
        traverse(root.right, ss, result);
    }
}

File: solutions/0129/1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int sum = 0;
        int d = 0;
        traverse(root, d, sum);
        return sum;
    }

    void traverse(TreeNode* root, int d, int& sum) {
        if (root == nullptr) {
            return;
        }
        int dd = d * 10 + root->val;
        if (root->left == nullptr && root->right == nullptr) {
            sum += dd;
        }
        traverse(root->left, dd, sum);
        traverse(root->right, dd, sum);
    }
};
